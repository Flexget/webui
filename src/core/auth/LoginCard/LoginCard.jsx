import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm } from 'redux-form';
import CardActions from '@material-ui/core/CardActions';
import TextField from 'common/TextField';
import {
  Card,
  ErrorMessage,
  LoginButton,
  Content,
  LoginField,
} from './styles';

export const LoginCard = ({ handleSubmit, errorStatus }) => (
  <Card>
    <form onSubmit={handleSubmit}>
      <Content>
        <ErrorMessage>
          {errorStatus.message}
        </ErrorMessage>
        <LoginField
          name="username"
          component={TextField}
          id="username"
          label="Username"
          autoFocus
        />
        <LoginField
          name="password"
          component={TextField}
          id="password"
          label="Password"
          type="Password"
        />
      </Content>
      <CardActions>
        <LoginButton type="submit">
          Login
        </LoginButton>
      </CardActions>
    </form>
  </Card>
);

LoginCard.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  errorStatus: PropTypes.object,
};

LoginCard.defaultProps = {
  errorStatus: {},
};

export default reduxForm({
  form: 'login',
})(LoginCard);
