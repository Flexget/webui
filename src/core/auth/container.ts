import { useState } from 'react';
import { createContainer } from 'unstated-next';

const useAuth = () => {
  const [isLoggedIn, setIsLoggedIn] = useState(false);

  const login = () => setIsLoggedIn(true);
  const logout = () => setIsLoggedIn(false);
  return [isLoggedIn, { login, logout }] as const;
};

export const AuthContainter = createContainer(useAuth);
