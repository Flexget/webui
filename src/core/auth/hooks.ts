export const useAuth = () => {
  const [isLoggedIn, setIsLoggedIn] = useState(false);

  let login = () => 
}
