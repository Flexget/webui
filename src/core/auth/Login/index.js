import { connect } from 'react-redux';
import { Constants } from 'core/version/state/actions';
import Login from './Login';

export function mapStateToProps({ auth, status }) {
  return {
    redirectToReferrer: !!auth.loggedIn,
    loading: !!status.loading[Constants.GET_VERSION],
  };
}

export default connect(mapStateToProps)(Login);
export { Login };
