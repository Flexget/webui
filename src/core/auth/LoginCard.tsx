import React, { FC } from 'react';
import { Form, Formik } from 'formik';
import { useContainer } from 'unstated-next';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import { CardContent, Card, Button, CardActions } from '@material-ui/core';
import { useFlexgetAPI } from 'core/api';
import { Method } from 'utils/fetch';
import TextField from 'common/inputs/formik/TextField';
import theme from 'core/theme';
import { AuthContainer } from './container';
import { LoginReq } from './types';

const card = css`
  max-width: 40rem;
  margin: 0 1rem;
  ${theme.breakpoints.up('sm')} {
    margin: 0 auto;
  }
`;

const field = css`
  padding-bottom: 1rem;
`;

const ErrorMessage = styled.div`
  color: ${theme.palette.error[500]};
  text-align: center;
  padding: 1rem;
`;

const content = css`
  display: flex;
  flex-direction: column;
`;

const LoginCard: FC = () => {
  const initialValues: LoginReq = { username: 'flexget', password: '' };
  const [, setLoggedIn] = useContainer(AuthContainer);

  const [{ error, loading }, login] = useFlexgetAPI('/auth/login', Method.Post);

  return (
    <Card css={card}>
      <Formik
        initialValues={initialValues}
        onSubmit={async (values, actions) => {
          const resp = await login(values);
          if (resp.ok) {
            actions.resetForm();
            setLoggedIn(true);
          }
        }}
      >
        <Form>
          <CardContent css={content}>
            <ErrorMessage>{error?.message}</ErrorMessage>
            <TextField css={field} name="username" id="username" label="Username" autoFocus />
            <TextField css={field} name="password" id="password" label="Password" type="Password" />
          </CardContent>
          <CardActions>
            <Button type="submit" fullWidth disabled={loading}>
              Login
            </Button>
          </CardActions>
        </Form>
      </Formik>
    </Card>
  );
};

export default LoginCard;
