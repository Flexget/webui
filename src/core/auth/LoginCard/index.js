import { connect } from 'react-redux';
import actions from 'core/auth/state/actions';
import LoginCard from './LoginCard';

export function mapStateToProps({ status }) {
  return {
    initialValues: {
      username: 'flexget',
    },
    errorStatus: status.error || {},
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onSubmit: ({ username, password }) => dispatch(actions.login.request(username, password)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(LoginCard);
export { LoginCard };
