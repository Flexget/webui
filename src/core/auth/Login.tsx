import React, { FC, useEffect } from 'react';
import { useContainer } from 'unstated-next';
import { Redirect, RouteComponentProps } from 'react-router-dom';
import { StaticContext } from 'react-router';
import { AuthContainer } from 'core/auth/hooks';
import { useFlexgetAPI } from 'core/api';
import LoginCard from './LoginCard';
import SplashScreen from './Splash';
import Logo from './Logo';

type Props = Partial<RouteComponentProps<{}, StaticContext, { from: Location }>>;

const LoginPage: FC<Props> = ({ location }) => {
  const { from } = location?.state ?? { from: { pathname: '/' } };
  const [loggedIn, setLoggedIn] = useContainer(AuthContainer);

  const [loginState, checkLogin] = useFlexgetAPI('/auth/verify');

  useEffect(() => {
    const fn = async () => {
      const resp = await checkLogin();
      if (resp.ok) {
        setLoggedIn(true);
      }
    };
    fn();
  }, [checkLogin, setLoggedIn]);

  if (loginState.loading) {
    return <SplashScreen />;
  }

  if (loggedIn) {
    return <Redirect to={from} />;
  }

  return (
    <div>
      <Logo />
      <LoginCard />
    </div>
  );
};

export default LoginPage;
