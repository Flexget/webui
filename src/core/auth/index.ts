import Login from './Login/index';

export default Login;
