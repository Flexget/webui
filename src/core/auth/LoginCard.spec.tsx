import React from 'react';
import { cleanup, fireEvent } from '@testing-library/react';
import fetchMock from 'fetch-mock';
import { AuthContainer } from 'core/auth/container';
import { renderWithWrapper } from 'utils/tests';
import LoginCard from './LoginCard';

const TestLogin = () => {
  const [loggedIn] = AuthContainer.useContainer();

  return (
    <>
      <div aria-label="login-state">{`${loggedIn}`}</div>
      <LoginCard />
    </>
  );
};

describe('core/Login/LoginCard', () => {
  beforeEach(() => {
    fetchMock
      .get('/api/tasks', [])
      .post('/api/auth/login', {})
      .catch();
  });

  afterEach(() => {
    cleanup();
    fetchMock.reset();
  });

  it('should call handleSubmit on submit', async () => {
    const { findByLabelText, getByRole } = renderWithWrapper(<TestLogin />);
    expect(await findByLabelText('login-state')).toHaveTextContent('false');

    const form = getByRole('form');
    fireEvent.submit(form);
    expect(await findByLabelText('login-state')).toHaveTextContent('true');
  });
});
