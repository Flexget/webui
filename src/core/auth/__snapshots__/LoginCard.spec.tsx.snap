// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/Login/LoginCard LoginCard renders correctly 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  css={[Function]}
>
  <form
    action="#"
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="MuiCardContent-root"
      css={
        Object {
          "map": undefined,
          "name": "1iyoj2o",
          "next": undefined,
          "styles": "
  display: flex;
  flex-direction: column;
",
          "toString": [Function],
        }
      }
    >
      <div
        css={[Function]}
      />
      <div
        className="MuiFormControl-root MuiTextField-root"
        css={
          Object {
            "map": undefined,
            "name": "1hq8d8z",
            "next": undefined,
            "styles": "
  padding-bottom: 1rem;
",
            "toString": [Function],
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
          data-shrink={true}
          htmlFor="username"
          id="username-label"
        >
          Username
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={true}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="username"
            name="username"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value="flexget"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root"
        css={
          Object {
            "map": undefined,
            "name": "1hq8d8z",
            "next": undefined,
            "styles": "
  padding-bottom: 1rem;
",
            "toString": [Function],
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink={false}
          htmlFor="password"
          id="password-label"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="Password"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Login
        </span>
      </button>
    </div>
  </form>
</div>
`;
